business/authentication/session_persistor.go

```diff
@@ -0,0 +1,335 @@
+	}
+
+	aesGcmNonce, err := util.CryptoRandomBytes(aesGcm.NonceSize())
+	if err != nil {
+		return fmt.Errorf("error when creating credentials - failed to generate random bytes: %w", err)
+	}
+
+	cipherSessionData := aesGcm.Seal(aesGcmNonce, aesGcmNonce, sDataJson, nil)
+	base64SessionData := base64.StdEncoding.EncodeToString(cipherSessionData)
+
+	// The base64SessionData holds what we want to store in browser cookies.
+	// It's time to set/send the browser cookies to persist the session.
+
+	// If the resulting session data is large, it may not fit in one cookie. So, the resulting
+	// session data is broken in chunks and multiple cookies are used, as is needed.
+	conf := config.Get()
+
+	sessionDataChunks := chunkString(base64SessionData, business.SessionCookieMaxSize)
+	for i, chunk := range sessionDataChunks {
+		var cookieName string
+		if i == 0 {
+			// Set a cookie with the regular cookie name with the first chunk of session data.
+			// Notice that an "-aes" suffix is being used in the cookie names. This is for backwards compatibility and
+			// is/was meant to be able to differentiate between a session using cookies holding encrypted data, and the older
+			// less secure sessions using cookies holding JWTs.
+			cookieName = config.TokenCookieName + "-aes"
+		} else {
+			// If there are more chunks of session data (usually because of larger tokens from the IdP),
+			// store the remainder data to numbered cookies.
+			cookieName = fmt.Sprintf("%s-aes-%d", config.TokenCookieName, i)
+		}
+
+		authCookie := http.Cookie{
+			Name:     cookieName,
+			Value:    chunk,
+			Expires:  expiresOn,
+			HttpOnly: true,
+			Path:     conf.Server.WebRoot,
+			SameSite: http.SameSiteStrictMode,
+		}
+		http.SetCookie(w, &authCookie)
+	}
+
+	if len(sessionDataChunks) > 1 {
+		// Set a cookie with the number of chunks of the session data.
+		// This is to protect against reading spurious chunks of data if there is
+		// any failure when killing the session or logging out.
+		chunksCookie := http.Cookie{
+			Name:     config.TokenCookieName + "-chunks",
+			Value:    strconv.Itoa(len(sessionDataChunks)),
+			Expires:  expiresOn,
+			HttpOnly: true,
+			Path:     conf.Server.WebRoot,
+			SameSite: http.SameSiteStrictMode,
+		}
+		http.SetCookie(w, &chunksCookie)
+	}
+
+	return nil
+}
+
+// ReadSession restores (decrypts) and returns the data that was persisted when using the CreateSession function.
+// If a payload is provided, the original data is parsed and stored in the payload argument. As part of restoring
+// the session, validation of expiration time is performed and no data is returned assuming the session is stale.
+// Also, it is verified that the currently configured authentication strategy is the same as when the session was
+// created.
+func (p CookieSessionPersistor) ReadSession(r *http.Request, w http.ResponseWriter, payload interface{}) (*sessionData, error) {
```

Comment:
It can be. In this particular case, I don't have preference of one over the other.

I can change it if you think it's better the pointer receiver.

Reply:
I wouldn't say pointer is better here, mostly just curious since I usually default to pointers for structs but I don't have a great reason for that other than cargo culting. `CookieSessionPersistor` is small and if it's going to be used concurrently then non-pointer might be better anyways.